### PR TITLE
Refactor MakeATangentVector by using VectorValues and fix testcase failure

### PR DESCRIPTION
**Pull request by jingwuOUO**
Refactor MakeATangentVector by using VectorValues and fix testcase failure

### CODE REVIEW

File: gtsam/sfm/ShonanAveraging.cpp [hunk 1]
```diff
@@ -643,20 +643,26 @@ bool ShonanAveraging<d>::checkOptimality(const Values &values) const {
...
   // Create a tangent direction xi with eigenvector segment v_i
   const size_t dimension = SOn::Dimension(p);
-  const auto v_i = v.segment<d>(d * i);
-  Vector xi = Vector::Zero(dimension);
-  double sign = pow(-1.0, round((p + 1) / 2) + 1);
-  for (size_t j = 0; j < d; j++) {
-    xi(j + p - d - 1) = sign * v_i(d - j - 1);
-    sign = -sign;
+  for (size_t i = 0; i < v.size() / d; i++) {
+    // Create a tangent direction xi with eigenvector segment v_i
```
Comment (dellaert) [inline gtsam/sfm/ShonanAveraging.cpp:654]: remove duplicate comment above

File: gtsam/sfm/tests/testShonanAveraging.cpp [hunk 2]
```diff
@@ -158,17 +158,17 @@ TEST(ShonanAveraging3, CheckWithEigen) {
...
   // Check Eigenvalue with slow Eigen version, converts matrix A to dense matrix!
-  const Matrix S = ShonanAveraging3::StiefelElementMatrix(Qstar3);
-  auto A = kShonan.computeA(S);
-  bool computeEigenvectors = false;
-  Eigen::EigenSolver<Matrix> eigenSolver(Matrix(A), computeEigenvectors);
-  auto lambdas = eigenSolver.eigenvalues().real();
-  double minEigenValue = lambdas(0);
-  for (int i = 1; i < lambdas.size(); i++)
-      minEigenValue = min(lambdas(i), minEigenValue);
+  // const Matrix S = ShonanAveraging3::StiefelElementMatrix(Qstar3);
```
Comment (dellaert) [inline gtsam/sfm/tests/testShonanAveraging.cpp:161]: Please re-enable commented out code

File: gtsam/sfm/ShonanAveraging.cpp [hunk 3]
```diff
@@ -643,20 +643,26 @@ bool ShonanAveraging<d>::checkOptimality(const Values &values) const {
 }
 
 /* ************************************************************************* */
-/// Create a tangent direction xi with eigenvector segment v_i
+/// Create a VectorValues with eigenvector v
 template <size_t d>
-Vector ShonanAveraging<d>::MakeATangentVector(size_t p, const Vector &v,
-                                              size_t i) {
+VectorValues ShonanAveraging<d>::MakeATangentVectorValues(size_t p,
```
Comment (dellaert) [inline gtsam/sfm/ShonanAveraging.cpp:648]: Let's rename to simply `TangentVectorValues`

File: gtsam/sfm/ShonanAveraging.cpp [hunk 4]
```diff
@@ -690,14 +696,8 @@ template <size_t d>
...
-  for (auto it : lifted.filter<SOn>()) {
-    // Create a tangent direction xi with eigenvector segment v_i
-    // Assumes key is 0-based integer
-    const Vector xi = MakeATangentVector(p, minEigenVector, it.key);
-    // Move the old value in the descent direction
-    it.value = it.value.retract(xi);
-  }
-  return lifted;
+  VectorValues delta = MakeATangentVectorValues(p, minEigenVector);
+  return lifted.retract(delta);
```
Comment (dellaert) [inline gtsam/sfm/ShonanAveraging.cpp:700]: nice !

File: gtsam/sfm/ShonanAveraging.cpp [hunk 5]
```diff
@@ -643,20 +643,26 @@ bool ShonanAveraging<d>::checkOptimality(const Values &values) const {
 }
 
 /* ************************************************************************* */
-/// Create a tangent direction xi with eigenvector segment v_i
+/// Create a VectorValues with eigenvector v
```
Comment (dellaert) [inline gtsam/sfm/ShonanAveraging.cpp:646]: No need to repeat docstring from header

File: gtsam/sfm/tests/testShonanAveraging.cpp [hunk 6]
```diff
@@ -121,7 +121,7 @@ TEST(ShonanAveraging3, tryOptimizingAt4) {
 }
 
 /* ************************************************************************* */
-TEST(ShonanAveraging3, MakeATangentVector) {
+TEST(ShonanAveraging3, MakeATangentVectorValues) {
```
Comment (dellaert) [inline gtsam/sfm/tests/testShonanAveraging.cpp:124]: Looking back on this, I don't think this was a great test, as the Hat thing is not needed. I created a new "before" test, please use this to make the new test, making sure the expected vectorts are exactly the same:
```c++
  Vector9 v;
  v << 1, 2, 3, 4, 5, 6, 7, 8, 9;
  Vector expected0(10), expected1(10), expected2(10);
  expected0 << 0, 3, -2, 1, 0, 0, 0, 0, 0, 0;
  expected1 << 0, 6, -5, 4, 0, 0, 0, 0, 0, 0;
  expected2 << 0, 9, -8, 7, 0, 0, 0, 0, 0, 0;
  const Vector xi_0 = ShonanAveraging3::MakeATangentVector(5, v, 0);
  const Vector xi_1 = ShonanAveraging3::MakeATangentVector(5, v, 1);
  const Vector xi_2 = ShonanAveraging3::MakeATangentVector(5, v, 2);
  EXPECT(assert_equal(expected0, xi_0));
  EXPECT(assert_equal(expected1, xi_1));
  EXPECT(assert_equal(expected2, xi_2));
```

File: gtsam/sfm/ShonanAveraging.cpp [hunk 7]
```diff
@@ -643,20 +643,26 @@ bool ShonanAveraging<d>::checkOptimality(const Values &values) const {
...
-  double sign = pow(-1.0, round((p + 1) / 2) + 1);
-  for (size_t j = 0; j < d; j++) {
-    xi(j + p - d - 1) = sign * v_i(d - j - 1);
-    sign = -sign;
+  for (size_t i = 0; i < v.size() / d; i++) {
+    // Create a tangent direction xi with eigenvector segment v_i
+    // Assumes key is 0-based integer
+    const auto v_i = v.segment<d>(d * i);
+    Vector xi = Vector::Zero(dimension);
+    double sign = pow(-1.0, round((p + 1) / 2) + 1);
```
Comment (dellaert) [inline gtsam/sfm/ShonanAveraging.cpp:658]: add
```c++
double sign0 = pow(-1.0, round((p + 1) / 2) + 1); 
```
before the for loop, and here initialize
```
double sign = sign0;
```

File: gtsam/sfm/tests/testShonanAveraging.cpp (hunk 2 above)
Comment (jingwuOUO) [reply gtsam/sfm/tests/testShonanAveraging.cpp:161]: Uncommented the code. But minEigenValue is never used, my suggestion is that maybe we can add a test like "EXPECT_DOUBLES_EQUAL(0, minEigenValue, 1e-11);" to make sure both the eigen value computed by Spectra and Eigen are in the same range?

**Pull request merged**